File: T1 Q6 - Key exchange/ECDH.cpp
#include <botan/auto_rng.h>
#include <botan/ec_group.h>
#include <botan/ecdh.h>
#include <botan/hex.h>
#include <botan/pubkey.h>

#include <iostream>

int main()
{
    Botan::AutoSeeded_RNG rng;

    // Define elliptic curve domain and key derivation function
    const auto ec_domain = Botan::EC_Group::from_name("secp521r1");
    const std::string key_derivation_function = "KDF2(SHA-256)";

    // Generate ECDH key pairs for two parties
    Botan::ECDH_PrivateKey private_key_a(rng, ec_domain);
    Botan::ECDH_PrivateKey private_key_b(rng, ec_domain);

    // Extract public keys for both parties
    const auto public_key_a = private_key_a.public_value();
    const auto public_key_b = private_key_b.public_value();

    // Create key agreement objects for both parties
    Botan::PK_Key_Agreement key_agreement_a(private_key_a, rng, key_derivation_function);
    Botan::PK_Key_Agreement key_agreement_b(private_key_b, rng, key_derivation_function);

    // Derive shared secrets
    const auto shared_secret_a = key_agreement_a.derive_key(32, public_key_b).bits_of();
    const auto shared_secret_b = key_agreement_b.derive_key(32, public_key_a).bits_of();

    // Verify if both shared secrets match
    if (shared_secret_a != shared_secret_b)
    {
        std::cout << "Key share failed!\n";
        return 1;
    }

    // Output the shared key
    std::cout << "Shared key: " << Botan::hex_encode(shared_secret_a) << "\n";
    std::cout << "Key share successful!\n";

    return 0;
}

File: code/T1 Q6 - Key exchange/ECDH.py
import botan3 as botan
import os

def main():
    # Create a random number generator instance
    rng = botan.RandomNumberGenerator()

    # Define elliptic curve domain and key derivation function
    ec_domain = 'secp521r1'
    key_derivation_function = 'HKDF(SHA-256)'

    # Generate ECDH key pairs for two parties
    private_key_a = botan.PrivateKey.create('ECDH', ec_domain, rng)
    private_key_b = botan.PrivateKey.create('ECDH', ec_domain, rng)

    # Create key agreement objects for both parties using their private keys
    key_agreement_a = botan.PKKeyAgreement(private_key_a, key_derivation_function)
    key_agreement_b = botan.PKKeyAgreement(private_key_b, key_derivation_function)

    # Generate a random salt for key derivation
    salt_length = 16  # Length of the salt in bytes
    salt = os.urandom(salt_length)

    # Derive shared secrets using the other's public key and salt
    shared_secret_a = key_agreement_a.agree(key_agreement_b.public_value(), 32, salt)
    shared_secret_b = key_agreement_b.agree(key_agreement_a.public_value(), 32, salt)

    # Verify if both derived shared secrets are identical
    if shared_secret_a != shared_secret_b:
        print("Key share failed!")
        return 1

    # Output the shared key in hexadecimal format
    print(f"Shared key:", shared_secret_a.hex())
    print("Key share successful!")

    return 0

if __name__ == "__main__":
    main()
